registrations/serializers.py

```diff
@@ -66,27 +71,82 @@ class Meta:
-        if kwargs["context"]["request"].data.get("event", None):
-            event_id = kwargs["context"]["request"].data["event"]
-            event = Event.objects.filter(id=event_id).select_related("publisher")
-            if len(event) == 0:
-                raise DRFPermissionDenied(_("No event with id {event_id}"))
+    publisher = serializers.SerializerMethodField()
+
+    event = JSONLDRelatedField(
+        serializer=EventSerializer,
+        many=False,
+        view_name="event-detail",
+        queryset=Event.objects.all(),
+    )
+
+    created_time = DateTimeField(
+        default_timezone=pytz.UTC, required=False, allow_null=True
+    )
+
+    last_modified_time = DateTimeField(
+        default_timezone=pytz.UTC, required=False, allow_null=True
+    )
+
+    created_by = serializers.StringRelatedField(required=False, allow_null=True)
+
+    last_modified_by = serializers.StringRelatedField(required=False, allow_null=True)
+
+    def __init__(self, instance=None, *args, **kwargs):
+        super().__init__(instance=instance, *args, **kwargs)
+
+        event_dict = kwargs["context"]["request"].data.get("event", None)
+
+        # Check user permissions if creating a new registration.
+        # LinkedEventsSerializer __init__ checks permissions in case of new registration
+        if not instance and isinstance(event_dict, dict) and "@id" in event_dict:
+            event_url = event_dict["@id"]
+            event_id = urllib.parse.unquote(event_url.rstrip("/").split("/")[-1])
             user = kwargs["context"]["user"]
-            if (
-                user.is_admin(event[0].publisher)
-                or kwargs["context"]["request"].method in SAFE_METHODS
-            ):
+
+            try:
+                event = Event.objects.get(pk=event_id)
+                error_message = _(f"User {user} cannot modify event {event}")
+
+                if not user.is_admin(event.publisher):
+                    raise DRFPermissionDenied(error_message)
+
+                if isinstance(user, ApiKeyUser):
+                    # allow updating only if the api key matches instance data source
+                    if event.data_source != user.data_source:
+                        raise DRFPermissionDenied(_(error_message))
+                else:
+                    # allow updating only if event data_source has user_editable_resources set to True
+                    if not event.is_user_editable_resources():
+                        raise DRFPermissionDenied(_(error_message))
+            except Event.DoesNotExist:
                 pass
+
+    def create(self, request, *args, **kwargs):
+        try:
+            instance = super().create(request, *args, **kwargs)
+        except IntegrityError as error:
+            if "duplicate key value violates unique constraint" in str(error):
+                raise serializers.ValidationError(
+                    {"event": _("Event already has a registration.")}
```

Comment:
Is this message correct? I assume an event can have multiple registrations. Or am I missing a context here? 🤔

Reply:
https://github.com/City-of-Helsinki/linkedevents/blob/d945861f0d0dceb543bdfc5407b64dda75c91bf6/registrations/models.py#L17
🙂

Reply:
Well... that settles that 😂